### ISSUE TITLE
USPS Web Tools API is now deprecated

### ISSUE THREAD

**Issue by herbdool** [backdrop-contrib/ubercart#510]
I was evaluating Ubercart for a client and noticed that the USPS Web Tools API is now deprecated: https://www.usps.com/business/web-tools-apis/welcome.htm. I can't see any posted date of when it will be shutdown. Maybe they're providing a long lead time.

The new API uses Oauth authorization so it would be a bit of work to get that working in Ubercart. Just based on that, I'm pretty sure that Ubercart is still using the old API.

**Comment by bugfolder** [comment]
> ...The new API uses Oauth authorization so it would be a bit of work to get that working in Ubercart. Just based on that, I'm pretty sure that Ubercart is still using the old API.

It is. My Ubercart site uses a heavily modified version of the uc_usps module, which also uses the old API.

It's a little reassuring that they haven't given a shutdown date, but we should probably start working on migrating to the new API (ideally by initially supporting both, but providing a configuration switch for administrators to switch when they've got their new credentials set up).

**Comment by oadaeh** [comment]
They have now given a date: https://postalpro.usps.com/node/13834

I was approached by a client today about this, because they received and email from the USPS. Here is the relevant part:

"As published in the February 6, 2025 [Industry Alert](https://postalpro.usps.com/node/13834), the Web Tools API platform will be retired on January 25, 2026. To avoid any service disruptions, we encourage you to migrate to the new USPS APIs as soon as possible.

The functional equivalents of the Web Tools API are available in the USPS API Catalogue: [USPS APIs Developer Portal](https://developers.usps.com/apis).

Additionally, please be aware that the Web Tools Label API retired on July 14, 2024. For Label API migration resources, visit the Announcements section at [Web Tools APIs | USPS](https://www.usps.com/business/web-tools-apis/welcome.htm)."

**Comment by bugfolder** [comment]
There is an open issue in the UC/D7 queue for UPS switching to OAuth 2.0, and one of the commenters posted a patch just a few weeks ago. I'll take a look and see if it can be a useful source for handling USPS's switch.

https://www.drupal.org/project/ubercart/issues/3425000#comment-15990503

@laryn, @argiepiano, @robertgarrigos, have any of you ever worked with OAuth integration? We should probably strive to get this into Ubercart by mid-2025 (with a switch for people to switch to their new credentials) and start putting deprecation notices into the status report.

**Comment by herbdool** [comment]
The patch and the zip file posted there appear to be for the D7 branch.

**Comment by bugfolder** [comment]
> The patch and the zip file posted there appear to be for the D7 branch.

Yes, so it should be relatively straightforward to port in https://github.com/backdrop-contrib/ubercart/issues/533. (Famous last words...)

**Comment by herbdool** [comment]
I won't be able to easily test it since I'm not in the US.

**Comment by bugfolder** [comment]
> I won't be able to easily test it since I'm not in the US.

I use UC with USPS for a site and so should be able to exercise it thoroughly. But I will first work on the UPS version as a warm-up.

**Comment by bugfolder** [comment]
Having now [updated a version of the UPS submodule](https://github.com/backdrop-contrib/ubercart/issues/533) (which works, at least, for rate lookup), and started to read the USPS dev docs, this is looking reasonably doable. I'm thinking that the way to do this is to implement BOTH "legacy" and "OAuth2" versions within the module, with a radio button switch between them, and leaving the module initially set to "legacy", but introducing a status report warning to switch to "OAuth2" as early as possible if someone is still on "legacy".

And then sometime after legacy EOL, we can strip out the legacy functionality.

**Comment by argiepiano** [comment]
> have any of you ever worked with OAuth integration

I haven't, but I could try to test a patch, probably after things slow down a bit for me (late April).

**Comment by bugfolder** [comment]
I'm now able to do successful rate lookups using the new API, but there's still a lot of coding to do because there are lots of changes. So, making progress.

**Comment by bugfolder** [comment]
Progress report: I have a nearly complete version of the module working locally, but want to do a bunch more testing before I create the PR. Soon, I hope.

**Comment by bugfolder** [comment]
@herbdool, @oadaeh, @argiepiano, @laryn: I have now posted a PR. This is not fully complete (I'm still having trouble getting insurance to quote), but the basic process works, and I would very much like feedback and testing.

I will note that it's not just a change from the old user ID to OAuth2, or the switch from XML to JSON, that's involved here; the data that is passed to USPS is different, and the return data is VERY different from before. There's not a one-to-one correspondence between old-style rate names and new-style rate names. And, in particular, one can no longer submit multiple packages within a single rate request;  one must make multiple requests on a per-package basis. I've tried to keep the behavior as close to the old behavior as possible, but it's not going to be a completely transparent upgrade. I've added a file, UPGRADING.md, to the module to try to help people with the process.

As one example: for a product with these parameters:

Shipping type=envelope, Package type=VARIABLE
0.5 oz, 9” x 5” x 0.1”

The old-style quote (with online rates) was this:

<img width="768" alt="Image" src="https://github.com/user-attachments/assets/22d61654-fd70-4ce7-a37c-b35157856049" />

The quote with the new API is this:

<img width="768" alt="Image" src="https://github.com/user-attachments/assets/175bbb11-1d1b-4478-98cf-b38cb41cda12" />

Note that old-style "Priority Mail" is equivalent to "Priority Mail Machinable Single-piece."

Please try this out (on a test site, of course—but you will have to set up OAuth2 credentials with USPS) and post feedback here.

**Comment by bugfolder** [comment]
I've now got insurance quoting included (when selected in settings). Please test.

**Comment by herbdool** [comment]
What are the steps to test? I suppose one must get a different kind of access key?

**Comment by bugfolder** [comment]
Yes. please read UPGRADING.md for a description of how to get the key.

**Comment by herbdool** [comment]
I'm going to partially test this. I'm waiting on the client for new credentials.

**Comment by herbdool** [comment]
@bugfolder I've reviewed the code, but I haven't fully tested since I don't have credentials. I've added comments in the PR.

**Comment by bugfolder** [comment]
Thanks @herbdool for the extensive comments! I'll go through them this weekend.

**Comment by bugfolder** [comment]
I've updated the PR in response to @herbdool's comments; anyone who can, please take a look and try it out.

As time goes by, I'm also thinking about timing of release. Since the old API will turn off on January 26, it seems like we should give users something like 3 months to (a) discover the update (definitely a new minor version for UC), (b) find a good time to upgrade their site and discover how the new USPS rate types alter their human workflows and respond accordingly. That suggests something like an end-of-October release at the latest. It would be great to have some testing on dev versions of real-world sites that use uc_usps before then to discover if there's anything major that needs changing.

**Comment by herbdool** [comment]
@bugfolder I see you've had some activity on this pr. I still haven't gotten a new ID from the client to test this. Hopefully before the old one is closed down.

**Comment by bugfolder** [comment]
@herbdool, @argiepiano, @laryn, it would be great to see some exercise of this on dev versions of a real store to get some real-world feedback.

**Comment by argiepiano** [comment]
Thanks, @bugfolder. Unfortunately I don't have any site that uses ubercart at the moment, and am a bit pressed with time to create  testing one (esp. since I need to create a USPS account etc.) Perhaps @herbdool has something he can use for testing?

**Comment by herbdool** [comment]
I've sent a reminder to our client who needs this, that they need to create a new key for testing.

**Comment by laryn** [comment]
I don't have a client using this currently, but my daughter is trying to set up a simple site to sell greeting cards, so it might be overkill for her needs but I'm going to test it out there. A few early notes:

- USPS settings have fields for `USPS client ID` and `USPS client secret ` but the USPS site refers to them as `Consumer Key` and `Consumer Secret`, respectively. That was a little confusing (because there is a separate login ID required for the USPS account) and eventually I just pasted them in to test and it seemed to work.
- Under USPS settings > Quote options, I hadn't changed anything and the vertical tabs summary says "Each in own" but the default selection is on "All products in one package" when I click to that tab, so there's a mismatch somewhere.

**Comment by bugfolder** [comment]
The vertical tab labels are fixed. (They were switched.)

The terms "client ID" and "client secret" are OAuth terms, and in fact are the terms used by the USPS API. But you are of course correct that users won't know that and don't need to know that, and it would be far clearer to make the names in UC match the names presented to the user in USPS. So I've changed those. Thanks! (Great to have this one identified before we release the module. You'll need to re-enter the values because the config variable names have also changed for consistency.)

**Comment by laryn** [comment]
I'm getting back to this and did an initial test of a checkout page. I have to dig deeper to see if I missed something in the configuration/setup, but when I try to get a shipping quote I see:

> 400 OASValidation OpenAPI-Spec-Validation-Domestic-Prices with resource oas://domestic-prices-v3.yaml: failed with reason: [ERROR - Instance failed to match all required schemas (matched only 1 out of 2): [/allOf/0: ECMA 262 regex ^d{5}(?:[-s]d{4})?$ does not match input string ]ERROR - [Path '/originZIPCode'] ECMA 262 regex ^d{5}(?:[-s]d{4})?$ does not match input string : []]

**Comment by bugfolder** [comment]
That error looks like the zip code was missing. (a) Did you supply one? (b) Does the "legacy" settings return a quote with the same settings? You can switch back and forth (but clear caches in between switching).

If a missing ZIP was the problem, the failure should be a bit more graceful than a cryptic error message.

**Comment by laryn** [comment]
@bugfolder Thanks -- I do see the reference to the Zip code, but I do have a Zip code in the address (both delivery and billing addresses). I hadn't filled in the store address/zip, but it gave the same error after I did that, too.

**Comment by bugfolder** [comment]
OK, I'll do some further testing. Tied up tomorrow, but should have a chance on Friday. (Of course I did a bunch of testing already, but you know the usual programmer's tendency to subconsiously avoid all corners of parameter space where bugs reside when testing their own code.)

**Comment by bugfolder** [comment]
> I hadn't filled in the store address/zip, but it gave the same error after I did that, too.

@Laryn, did you do that for the product's pickup address? I can recreate this error message, but only if the product pickup address is missing its zip code.

If the origin zip code is missing (from the store or from the individual product), the lookup has to return an error of some sort. The only question would be do we present the full error to the end user, or just give something generic like "there was an error." I favor the former, because the USPS error message will give a hint as to what the problem is (missing originZIPCode), and that's something that only the site admin can address, so we'd want the end user to report the full error message.

Per my comments above, next month is when I'd hope to release this, to give site admins 3 months to discover the update and figure out what their new settings need to be to keep their store running.

**Comment by oadaeh** [comment]
Thanks for taking this on, @bugfolder.

My client has reached back out to me, so I hope to be testing this over the next week or two to see what is up and how well it works.
I assume I can just have them follow the D7 instructions, here: https://www.drupal.org/node/1308256 to have them create a new user ID with USPS, so that's what I'm going to tell them.

Regarding your error reporting question, I also would prefer the former, but if you don't want to avoid freaking out the clients, you could record the full error in watchdog, but display an abbreviated one on the page with a note to look in the logs for more information.

**Comment by bugfolder** [comment]
@oadaeh, make sure they read the UPGRADING.md file in the module as well.

**Comment by oadaeh** [comment]
Oh, all the updates are still only in the PR. Thanks for that, because I was wondering why I couldn't find anything newer than "years ago." 😅

**Comment by herbdool** [comment]
I think we'll finally be able to test this PR. Our client finally responded once their shipping quotes stopped working entirely. When I look at the debugging I see:

> Error: Your Web Tools API service has been disrupted because we have been unable to contact you about migrating to the newest version of our API software. Please fill out the following Migration Feedback Form to begin upgrading and restoring your service: https://forms.osi.office365.us/r/n6njkKcCaz.  After completing the form, please contact webtools@usps.gov. Thank you for your attention to this matter.

**Comment by herbdool** [comment]
@bugfolder in testing it appears to be mostly working. I was getting some quotes at first. But sometimes it's not working and there's no debug info. I put in a `debug($rate_info)` to see if any quotes were coming back. Certainly they were, but it seems like there are options which are not even provided in the new services settings, such as:

- LMNS: Library Mail Nonstandard Single-piece
- MMNS: Media Mail Nonstandard Single-piece

Note, that I had kept it on the default Commercial price type. Though it doesn't seem to make a difference in what services are offered.

Now, those are not strictly causing any error. Instead, I'm wondering if a debug message could be shown (if debugging is enabled) if *no* services match?

**Comment by bugfolder** [comment]
> I'm wondering if a debug message could be shown (if debugging is enabled) if no services match?

That's a very good idea. I'll add that. (Also will add those two services.) I was unable to find a description of all possible services, so formed the list by just doing a slew of varying quotes. Can you describe the product attributes you used that gave those quote values?

It might also be worthwhile to do a check to see if quotes were being returned that aren't in the official list (whether selected or not) and put that, too, in the debug log.

**Comment by herbdool** [comment]
Another issue. I seem to be having a problem if there's more than one separate item in the checkout. It can handle more than one of the same kind, but not different. And `debug($rate_info)` is returning an empty array. (This is when selecting "All products in one package").

**Comment by herbdool** [comment]
>Can you describe the product attributes you used that gave those quote values?

11x10x10 inches, 20 pounds

**Comment by bugfolder** [comment]
>  I seem to be having a problem if there's more than one [different] separate item in the checkout.

Do you have "all in one" or "each in own" checked in the settings? And what are the attributes of the products?

And what does the legacy checkout return for rates for this scenario?

**Comment by herbdool** [comment]
Can't use the legacy right now. The client's old ID is still blocked.

**Comment by herbdool** [comment]
I had updated my previous comment that it was set to "All products in one package". It works when set to each in own.

**Comment by herbdool** [comment]
When using "All in one" and there are multiple products in the cart, `$packages = _uc_usps_package_products($products, $addresses);` looks like this:

```
array (
  0 =>
  array (
    0 =>
    (object) array(
       'price' => 99.85000000000001,
       'qty' => 1,
       'pounds' => 5.0,
       'ounces' => 6.400000000000006,
       'weight' => 5.4,
       'container' => 'VARIABLE',
       'size' => 'REGULAR',
       'machinable' => true,
       'length' => 0,
       'width' => 0,
       'height' => 0,
       'girth' => 0,
    ),
  ),
);
```

The two products did have length, width and height set but it's not showing up here.

If there's only one product in the cart, then the `$packages` array includes the `UcOrderProduct` objects with the dimensions intact. And similarly when using "Each in own" option.

**Comment by herbdool** [comment]
The odd thing is how the Calculated Shipping Cost box returns no rates (all in one, 2 products), but that it still adds "Priority Mail" to the sub-total.

<img width="808" height="504" alt="Image" src="https://github.com/user-attachments/assets/42e704a8-4583-40b7-8701-6bfb5026ea54" />

The error message is the one I was trying out.

**Comment by bugfolder** [comment]
> When using "All in one" and there are multiple products in the cart, $packages = _uc_usps_package_products($products, $addresses); looks like this:...

Yes, that was what I was suspecting. The packaging code (`_uc_usps_package_products ()`) is not significantly modified by this PR, meaning that this would also be the outcome with the legacy lookup as well.

And the broader problem there is that packaging diverse products into a single package can be a tough problem (like, it's NP-hard), so the original Ubercart basically punted on the problem. In my main Ubercart site, which uses a heavily modified version of `uc_usps`, the products are flat and stackable, so we can actually handle this. That could be an optional change for UC in the future. But my goal here was to just get UC to work on any scenario where it previously worked, and not try to get it to handle the scenarios where it didn't previously work (a much bigger chore).

**Comment by herbdool** [comment]
Yeah, I just noticed that as well, that UC didn't do anything with the dimensions other than weight. Makes sense in a way. I was wondering how it would calculate how to stack multiple items with different dimensions. So that probably didn't work before either. Or maybe USPS would still return a rate based on only the weight?

**Comment by herbdool** [comment]
Maybe the legacy did something with this `$product->usps['container'] = isset($temp->usps['container']) ? $temp->usps['container'] : 'VARIABLE';` and now nothing?

**Comment by herbdool** [comment]
In the meanwhile, I'm wondering if it's less confusing for the customer if "each in own" is used, since there's no way to prevent multiple items in a cart.

**Comment by herbdool** [comment]
No, the `container` doesn't make a difference. Now I noticed where it was set on the product. Still figuring some things out with Ubercart.

**Comment by herbdool** [comment]
I put suggestions on the PR which you could commit if that makes it easier @bugfolder.

**Comment by bugfolder** [comment]
Thanks, @herbdool; I've added the exterior services you put into the PR (thanks) and incorporated your other suggestions (though in a slightly different way, making us of the Ubercart debugging and logging settings in the Basic Settings section of uc_quote configuration).

**Comment by herbdool** [comment]
@bugfolder I had added a suggestion for LMMS and MMMS to uc_usps_newapi_dom_parcel_all_services() but forgot to add it to the default settings. So a couple more small suggestions.

**Comment by bugfolder** [comment]
Merged.

Also, in the last round of updates, I verified that all-in-one with two dissimilar packages failed to return _any_ quotes using the legacy settings and left package dimensions at zero. I added code to `_uc_usps_package_products()` that computed package dimensions under all-in-one only for newapi, getting dimensions by simply stacking up products. This now returns some quotes, but not all the services I would have expected; not entirely sure why yet.

BTW, I will be disconnected from internet 9/27 through 10/11 so won't be able to work on this during that time. Still would be nice to get the release out in October.

**Comment by herbdool** [comment]
We're using the patch in production now so getting some live testing. Currently it's set to "each in own" but maybe will get your latest version and try "all in one". We'll try it out and hopefully have enough testing to release it soon.

**Comment by bugfolder** [comment]
Please update (again). I've improved the debugging information (pretty-printing the JSON) and fixed a bug I found affecting the all-in-one quoting (which stacks the products in the package).

**Comment by herbdool** [comment]
Will do. I'll check again next week when I'm back at the computer.

**Comment by bugfolder** [comment]
If you find further problematic behavior, please, if at all possible, switch back to legacy and compare the two behaviors. 

Of course we want things to work for new installs, but I expect the urgency and pain problems to be heightened for existing legacy sites that upgrade and then find things not working for them. And some number of legacy sites are only going to upgrade after the USPS legacy API shuts down, their store mysteriously stops working, and they're losing sales every hour until they get a working upgrade in place.

**Comment by bugfolder** [comment]
I've done some more testing and bug-fixing, added a few more possible services to be returned. (Also added a detection scheme: when debugging is turned on, if USPS returns any services that we don't account for, a warning is put into watchdog with the missing services so that they can be added.)

I'd originally hoped to release this in October to give people 3 months to adjust before the old API goes away. That date has obviously slipped. The date of turn-off of the old API, January 25, 2026,  hasn't, of course, so I am inclined to go ahead and release the next version of Ubercart with this module and let the chips fall where they may. Any objections, @herbdool, @argiepiano, @laryn (or anyone else)?

**Comment by argiepiano** [comment]
@bugfolder, sorry I haven't been able to test all this - I am currently not using Ubercart and don't have a USPS account etc. I think the plan to go ahead and release the next version makes sense.

**Comment by laryn** [comment]
No objections from me.

**Comment by herbdool** [comment]
@bugfolder okay by me. I haven't heard any problems reported from the one site using this patch.
